### PR TITLE
Alignment tracking in coin-cell battery robot with camera

### PR DESCRIPTION
**Pull request by linasofie**

### CODE REVIEW

File: aurora_robot_tools/camera/process_image.py [hunk 1]
```diff
@@ -0,0 +1,570 @@
...
+        bottom_half_sorted = sorted(higher_y_group, key=lambda x: x[0])
+        # Arrange in desired order: upper left, upper right, lower right, lower left
+        centers_sorted = np.float32([top_half_sorted[0], top_half_sorted[-1],
+                                     bottom_half_sorted[-1], bottom_half_sorted[0]])
+        # Transform Perspective
+        M = cv2.getPerspectiveTransform(centers_sorted, pts2) # transformation matrix
+        return M
+
+    def _transform_split(self, img: np.array, m: np.array, filename: str) -> dict[np.array]:
+        """ Takes image and transformation matrix and returns transformed image.
```
Comment (g-kimbell) [inline aurora_robot_tools/camera/process_image.py:288]:
```suggestion
        """Transform and crop an image to give one sub-image per press position.
```

File: aurora_robot_tools/camera/process_image.py [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+            bottom_right_y = (c[1] + 2*self.offset_mm) * self.mm_to_pixel
+            bottom_right_x = (c[0] + 2*self.offset_mm) * self.mm_to_pixel
+            top_left_y = max(bottom_right_y - 2*self.offset_mm*self.mm_to_pixel, 0)
+            top_left_x = max(bottom_right_x - 2*self.offset_mm*self.mm_to_pixel, 0)
+            cropped_image = transformed_image[top_left_y:bottom_right_y, top_left_x:bottom_right_x]
+            cropped_images[i+1] = cropped_image
+        return cropped_images
+
+    def _thickness_correction(self, p: int, s: int, center: tuple) -> tuple:
+        """ Correcting position of center with thickness of parts.
```
Comment (g-kimbell) [inline aurora_robot_tools/camera/process_image.py:319]:
```suggestion
        """Correct center position for component height.
```

File: aurora_robot_tools/camera/process_image.py [hunk 3]
```diff
@@ -0,0 +1,570 @@
...
+        # Save the image with detected ellipses
+        cv2.imwrite(self.path + f"/transformed/{filename.split(".")[0]}.jpg", transformed_image)
+        # Crop the image
+        cropped_images = {}
+        for i, c in enumerate(self.press_position):
+            # set zero in case it gives a negative number
+            bottom_right_y = (c[1] + 2*self.offset_mm) * self.mm_to_pixel
+            bottom_right_x = (c[0] + 2*self.offset_mm) * self.mm_to_pixel
+            top_left_y = max(bottom_right_y - 2*self.offset_mm*self.mm_to_pixel, 0)
+            top_left_x = max(bottom_right_x - 2*self.offset_mm*self.mm_to_pixel, 0)
```
Comment (g-kimbell) [inline aurora_robot_tools/camera/process_image.py:313]: Why do you use 2x offset?

File: aurora_robot_tools/camera/process_image.py [hunk 4]
```diff
@@ -0,0 +1,570 @@
...
+        # store in data frame
+        df["dx_mm_corr"] = x_corrected
+        df["dy_mm_corr"] = y_corrected
+        df["dz_mm_corr"] = np.sqrt(df["dx_mm_corr"]**2 + df["dy_mm_corr"]**2).round(5)
+        self.df = df
+        return df
+
+    def save(self) -> pd.DataFrame:
+        """ Saves data with all coordinates, radius and alignment.
+        """
```
Comment (g-kimbell) [inline aurora_robot_tools/camera/process_image.py:505]:
```suggestion
        """Save data with all coordinates, radii and alignments."""
```

File: aurora_robot_tools/camera/process_image.py [hunk 5]
```diff
@@ -0,0 +1,570 @@
...
+        data_dir_data = os.path.join(self.path, "data")
+        self.df = self.df.drop(columns=["array"]) # save without image array
+        if not os.path.exists(data_dir_data):
+            os.makedirs(data_dir_data)
+        with pd.ExcelWriter(os.path.join(data_dir_data, "data.xlsx")) as writer:
+            self.df.to_excel(writer, sheet_name='coordinates', index=False)
+
+        return self.df
+
+#%% RUN CODE
```
Comment (g-kimbell) [inline aurora_robot_tools/camera/process_image.py:548]: Remove cell-mode comments - these are IDE-dependent

File: aurora_robot_tools/camera/process_image.py [hunk 6]
```diff
@@ -0,0 +1,570 @@
...
+import cv2
+import re
+import json
+import numpy as np
+import pandas as pd
+from PIL import Image
+from scipy import signal
+import sqlite3
+
+#%% FUNCTIONS
```
Comment (g-kimbell) [inline aurora_robot_tools/camera/process_image.py:63]: Remove cell-mode comments, these are IDE-dependent

File: aurora_robot_tools/camera/process_image.py [hunk 7]
```diff
@@ -0,0 +1,570 @@
...
+        processed_image = image # no preprossessing
+    return processed_image
+
+#%% CLASS
+
+class ProcessImages:
+    def __init__(self, path):
+        # TRANSFORMATION ---------------------------------------------------------------------------
+        self.path = path # path to images
+        self.run_ID = self.path.split("/")[-1] # get run_ID from path
```
[the remark this answers is not in the document]
Comment (linasofie) [reply aurora_robot_tools/camera/process_image.py:200]: I changed it to Pathlib, can you check if it is the way you thought of it?

File: aurora_robot_tools/camera/process_image.py (hunk 3 above)
Comment (linasofie) [reply aurora_robot_tools/camera/process_image.py:313]: True that does not make sense. I changed the offset itself from 20 to 40 and deleted the multiply by 2.

**Comment by g-kimbell** [comment]
Squashed and merged, well done on everything!

**Pull request merged**